Added an overloaded version of toHost() and toAccelerator() that takes a tensor shape, to be used for XLA/TPU compilation

mhong opened the pull request.

Commit:
Added an overloaded version of toHost() and toAccelerator() that takes a tensor
shape, to be used for XLA/TPU compilation.

mhong commented:
@swift-ci please test tensorflow

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
+    // If the `self` tensor resides on TPU, the shape is specified on that
+    // device first, before outfeeding the tensor to CPU, a required step for
+    // sending the tensor to the host.
+    let tensor : Tensor<Scalar> = #tfop("Identity", self, __shapes: [shape])
```
Remove space before `:`.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
+    // device first, before outfeeding the tensor to CPU, a required step for
+    // sending the tensor to the host.
+    let tensor : Tensor<Scalar> = #tfop("Identity", self, __shapes: [shape])
+    return Tensor(handle: _TFReceive(tensor.handle))
```
Just `tensor.toHost()` would do it.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
+  ///   must specify the tensor shape as required by XLA compilation.
+  @inlinable @inline(__always)
+  func toAccelerator(shape: TensorShape) -> Tensor {
+    let tensor = Tensor(handle: _TFSend(handle))
```
Just `let tensor = toAccelerator()`

Commit:
Addressed comments from @rxwei.

rxwei reviewed (approved):
Nice!

lattner commented on lib/SILOptimizer/Mandatory/TFLowerGraph.cpp:
```diff
@@ -1230,13 +1230,13 @@ void TFGraphLowering::addTPUEnqueueOp(BuiltinInst *inst, bool isInfeed,
       internalError(
           getUserSourceLocation(inst->getDebugLocation()),
-          "TPU outfeed enqueue support enqueuing a single tensor -- did "
+          "TPU outfeed enqueue supports enqueuing a single tensor -- did "
```
Random question, but why does outfeed need a shape?  It has an operand which should have a shape, can't it get it from there?

mhong replied:
Indeed this is not required for infeed enqueue, and I removed the TFLowerGraph logic for its shapes attr. For outfeed enqueue, there is not even a shapes attr (https://b.corp.google.com/issues/110538524).

lattner commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
+    return tensor.toHost()
+  }
+
+  // TODO: Unify these two versions of toAccelerator() by making `shape` above
```
I actually like the way you have it with two overloads, I don't see a need to merge these.

mhong replied:
k, removed the TODOs here.

lattner reviewed (approved):
Looks great!

Commit:
Addressed coents from @lattner. When lowering to TPU infeed enqueue op,
make the shapes attr optional. If specified, it's only used for error checking,
so the enqueue ops in general know about the tensor shape of their input tensor.

Note the shapes attr for infeed/output dequeue ops is still required.

mhong commented:
Thank you both for the review!
